hedera-node/docs/design/services/smart-contract-service/frictionless-airdrops-system-contracts.md

```diff
@@ -0,0 +1,164 @@
+# Frictionless Airdrops via System Contracts
+
+## Purpose
+
+[HIP-904](https://hips.hedera.com/hip/hip-904) introduced the Frictionless Airdrops feature for fungible and non-fungible tokens.
+This document will define the architecture and implementation of the `airdropToken`, `claimAirdrops`, `cancelAirdrops` and `rejectTokens` smart contract functions
+and their respective redirect function calls (`cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFT`, `rejectTokensNFT`, `setAutomaticAssociations`) that will extend the capabilities of the Hedera Smart Contract Service (HSCS) to support frictionless airdrops.
+
+## References
+
+[HIP-904](https://hips.hedera.com/hip/hip-904) - HIP that introduces the frictionless airdrops.
+
+## Goals
+
+- Expose `airdropToken`, `claimAirdrops`, `cancelAirdrops` and `rejectTokens` as new functions in the Hedera Token Service Smart Contract.
+- Expose `cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFt`, `rejectTokensNFT`, `setAutomaticAssociations` as new functions in the proxy redirect token facade contract IHRC.
+- Implement the needed HTS system contract classes to support the new functions.
+
+## Non Goals
+
+- The implementation of the HAPI operation, as it is already an existing feature.
+
+## Architecture
+
+The architecture for the frictionless airdrops follows the existing framework defined for handling all calls to the HederaTokenService system contract in the modularization services and is described in more detail in the Implementation section below.
+
+## Implementation
+
+### New Solidity Structures
+
+We will introduce the following new structures to support the new functionality:
+
+`PendingAirdrop` - A struct that represents a pending airdrop request.
+
+```solidity
+struct PendingAirdrop {
+    address sender;
+    address receiver;
+
+    address token;
+    bool isNft;
+    int64 serial;
+}
+```
+
+`PendingAirdropRecord` - A struct that represents a pending airdrop record.
+
+```solidity
+struct PendingAirdropRecord {
+    PendingAirdrop pendingAirdrop;
+    uint64 amount;
+}
+```
+
+`NftId` - A struct that represents the Nft serials to be rejected.
+
+```solidity
+struct NftId {
```

Comment:
Can this struct be used in other places than just related to the airdrop functionality?

Reply:
I think we can incorporate it with functions that manipulate/read data for multiple NFTs with multiple serials.